lib/internal/modules/esm/translators.js

```diff
@@ -159,16 +165,18 @@ translators.set('module', async function moduleStrategy(url, source, isMain) {
     importModuleDynamically,
   });
   return module;
-});
+}
+translators.set('module', moduleStrategy);
 
 /**
- * Provide a more informative error for CommonJS imports.
- * @param {Error | any} err
+ * Provide a more informative error for CommonJS erroring because of ESM syntax,
+ * when we aren't retrying CommonJS modules as ES modules.
+ * @param {Error | unknown} err
  * @param {string} [content] Content of the file, if known.
  * @param {string} [filename] Useful only if `content` is unknown.
  */
 function enrichCJSError(err, content, filename) {
-  if (err != null && ObjectGetPrototypeOf(err) === SyntaxErrorPrototype &&
+  if (!detectModule && err != null && ObjectGetPrototypeOf(err) === SyntaxErrorPrototype &&
```

Comment:
I was trying to go from least expensive call to most, to error after the earliest cheapest call. Wouldn't checking the flag be cheaper than checking the error?

And yes I think the new detect error function is better than the old. I can open a separate PR to replace the old one rather than have two.

Reply:
We already check for SyntaxErrors unconditionally in `enrichCJSError` (that uses `hasEsmSyntax`) when the module being compiled is the entry point. I think we can simply change all the caller of `enrichCJSError` to consider detection before emit the warning. So before it's `if (hasEsmSyntax(error)) emitWarning(..)`, now it's `if (isModuleSyntaxError(error)) { if (detectModule) {..} else { emitWarning(); } }`

Reply:
Actually `Module.prototype._compile` does not return anything. It is not even just compiling modules, but also actually loading them. I think we can just do the detection-reload either there or in `wrapSafe` (well `wrapSafe` sounds like a weird place to do it, although this is shared by embedder and SEA so by changing it you can affect those use cases too, or it can be refactored a bit to create a helper that sounds more sensible.).

Reply:
We can't do the retry within the CommonJS loader because that loader can't load ESM. That's why we need to carry the exception all the way back up to this level.

Reply:
Why can't it load ESM? Doesn't it already do that in the `importModuleDynamically` callback? It doesn't seem too different to start the ESM loader when `import()` is encountered compared to starting the ESM loader when e.g. `import` is encountered.

Reply:
> Doesn’t it already do that in the `importModuleDynamically` callback?

I think it calls out to the ESM loader to load that dependency, as in it doesn’t start going down the `require` flow for a file that’s referenced via `import()`, whereas by the time we’ve compiled the current file we’re already deep into the `require` flow and we somehow need to unwind all that. But if I’m wrong, sure, that would be great. You’re right in that however `import()` works is probably the basis for how to make the CJS loader tell the ESM loader to handle a particular module.